src/bson_mapper/test/archiver_test.cpp

```diff
@@ -0,0 +1,462 @@
+ */
+bool file_cmp(std::string filename1, std::string filename2) {
+    std::ifstream file1(filename1, std::ios_base::binary);
+    std::ifstream file2(filename2, std::ios_base::binary);
+
+    file1.seekg(0, file1.end);
+    file2.seekg(0, file2.end);
+
+    size_t file1_length = file1.tellg();
+    size_t file2_length = file2.tellg();
+
+    file1.seekg(0, file1.beg);
+    file2.seekg(0, file2.beg);
+
+    if (file1_length != file2_length) {
+        return false;
+    }
+
+    char file1buf[file1_length], file2buf[file2_length];
+
+    file1.read(file1buf, file1_length);
+    file2.read(file2buf, file2_length);
+
+    return std::memcmp(file1buf, file2buf, file1_length) == 0;
+}
+
+/**
+ * Counts the number of keys in a particular document view.
+ *
+ * @param v
+ *      the bsoncxx::document::view to count keys in.
+ *
+ * @return the number of keys in the view.
+ */
+size_t countKeys(bsoncxx::document::view v) {
+    size_t count = 0;
+    for (auto elem = v.begin(); elem != v.end(); ++elem) {
+        ++count;
+        // DEBUG: std::cout << " key: " << elem->key() << "\n";
+    }
+    return count;
+}
+
+struct DataA {
+    int32_t x, y;
+    double z;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(CEREAL_NVP(x), CEREAL_NVP(y), CEREAL_NVP(z));
+    }
+
+    bool operator==(const DataA& other) {
+        return x == other.x && y == other.y && z == other.z;
+    }
+};
+
+struct BadDataA {
+    int32_t x, y;
+    double z;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(x, y, z);
+    }
+};
+
+struct DataB {
+    int64_t a, b;
+    DataA m;
+    std::vector<DataA> arr;
+    std::string s;
+    std::chrono::system_clock::time_point tp;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(CEREAL_NVP(a), CEREAL_NVP(b), CEREAL_NVP(m), CEREAL_NVP(arr), CEREAL_NVP(s),
+           CEREAL_NVP(tp));
+    }
+};
+
+int someInt = 42;
+
+struct DataC : public cereal::UnderlyingBSONDataBase {
+    bsoncxx::types::b_double test_double;
+    bsoncxx::types::b_utf8 test_utf8;
+    bsoncxx::types::b_document test_doc;
+    bsoncxx::types::b_array test_arr;
+    bsoncxx::types::b_binary test_bin;
+    bsoncxx::types::b_oid test_oid;
+    bsoncxx::oid test_oid2;
+    bsoncxx::types::b_bool test_bool;
+    bsoncxx::types::b_date test_date;
+    bsoncxx::types::b_int32 test_int32;
+    bsoncxx::types::b_int64 test_int64;
+
+    template <class Archive>
+    void serialize(Archive& ar) {
+        ar(CEREAL_NVP(test_double), CEREAL_NVP(test_utf8), CEREAL_NVP(test_doc),
+           CEREAL_NVP(test_arr), CEREAL_NVP(test_bin), CEREAL_NVP(test_oid), CEREAL_NVP(test_oid2),
+           CEREAL_NVP(test_bool), CEREAL_NVP(test_date), CEREAL_NVP(test_int32),
+           CEREAL_NVP(test_int64));
+    }
+
+    DataC()
+        : test_double{1.4},
+          test_utf8("cxx-odm"),
+          test_bin{bsoncxx::binary_sub_type::k_binary, sizeof(int),
```

Comment:
Not sure if you intend to get different serialization on big/little endian (or if you're comparing this with a fixed point like a binary file).

You might prefer just having some bytes to put in the binary to side skip the issue.  I.e.

uint8_t someBytes[] = { 1, 2, 3, 4 };